### ISSUE TITLE
filteredOn methods don't retain provided description or comparators

### ISSUE THREAD

**Issue by joel-costigliola** [joel-costigliola/assertj-core#1203]
#### Summary

`filteredOn` don't retain provided description.

#### failing test

```java
  @Test
  public void should_keep_existing_description() {
    // WHEN
    Throwable error = catchThrowable(() -> assertThat(employees).as("oops").filteredOn("age", 800).isEmpty());
    // THEN
    assertThat(error).hasMessageContaining("oops");
  }
```
#### solution

Chaining the underlying with `.as(info.description())` as in `extracting` methods.
 
It turns out that the code does not compile because of incompatible return type, I think this can be fixed by making `filteredOn` to return `SELF` instead of the `AbstractListAssert`, that makes sense as the initial iterable is filtered, its element type is not changes so we are dealing with the same assertion type.

Returning `SELF` instead of `AbstractListAssert` is a minor breaking change.

**Comment by joel-costigliola** [comment]
Same goes for comparators, for this test fails:

```java
@Test
public void should_retain_element_comporator() {
  // GIVEN
  Set<String> set = newLinkedHashSet("Foo", "Bar");
  // SUCCEEDS
  assertThat(set).usingElementComparator(CaseInsensitiveStringComparator.instance)
                 .contains("foo", "bar");
  // FAIL
  assertThat(set).usingElementComparator(CaseInsensitiveStringComparator.instance)
                 .filteredOn($ -> true)
                 .contains("foo", "bar");
}
```
The idea here is propagate any assertions context, in this case it means these fields:

- iterables
- objects
- info
- comparatorsByType
- comparatorsForElementPropertyOrFieldNames
- comparatorsForElementPropertyOrFieldTypes

**Comment by joel-costigliola** [comment]
Since `info` is a final field, either make it non final or better just propagate any  fields:
- overridingErrorMessage
- description (a copy of)
- representation

**Comment by joel-costigliola** [comment]
@raphw in the `intercept` code below that is used by soft assertions, `Callable` is a call to an assertion method (on an assert instance) that changes the object under test, like [`filteredOn`](http://joel-costigliola.github.io/assertj/core-8/api/org/assertj/core/api/AbstractIterableAssert.html#filteredOn-java.util.function.Predicate-). Here we create another proxy to make sure the new assertion is still "soft" proxied. 

What I'm trying to do is to propagate assertion information like the [assertion description](http://joel-costigliola.github.io/assertj/core-8/api/org/assertj/core/api/Descriptable.html#as-java.lang.String-java.lang.Object...-), in order to do so I need to access the original `Assert` instance   - the one hidden inside `Callable` parameter that models an assertion method call - and copy it to the `newProxyAssert`.

```java
@RuntimeType
public AbstractAssert<?, ?> intercept(@SuperCall Callable<AbstractAssert<?, ?>> proxy) throws Exception {
    // proxy is a call to an AbstractAssert subclass method returning an AbstractAssert
    Object result = proxy.call();
    Object newProxy = proxies.create(result.getClass(), actualClass(result), actual(result));
    AbstractAssert<?, ?> newProxyAssert = (AbstractAssert<?, ?>) newProxy;
    // this raises a ClassCastException as proxy is not an AbstractAssert !
    return newProxyAssert.withAssertionInfo(((AbstractAssert<?, ?>) proxy).info);
  }
```

Is there a way to access the assert instance instance that is being called on ?

I see in my debugger that such information exists in `proxy.argument0.info` (`info` is the field containing the assertion info to propagate).
Oh and `proxy` looks like `org.assertj.core.api.ProxyableIterableAssert$ByteBuddy$quM8O42u$auxiliary$oxl7YuUo@146dfe6` if that helps. 

I'm tempted to use reflection to sneakily access `argument0.info` but it feels really hacky!
Am I misusing byte-buddy ?  Is there a more elegant solution ? Maybe with `Advice` (which I haven't had a look at) ?

As a final note, this fix is a nice to have but not critical, so that's ok if it can't be addressed, it is still possible to set a description but it needs to be done after calling `filteredOn` as illustrated below (no great but ok):

```java
// fails as expected and foo (test description) appears in the error message
softly.assertThat(names)
          .filteredOn(name -> name.first.startsWith("Jo"))
          .as("foo")
          .hasSize(123);

// fails as expected but foo does not appear in the error message, 
// foo (test description) is lost after calling the proxied filteredOn
softly.assertThat(names)
          .as("foo")
          .filteredOn(name -> name.first.startsWith("Jo"))
          .hasSize(123);
```

**Comment by raphw** [comment]
You can use `@This` to get the current instance's reference. This is what you need, right?

**Comment by joel-costigliola** [comment]
Fixed in 0a0e28014978f62961cf50823e171a53c010d2d7
